### PR TITLE
Add dmd as a lib callbacks for statement semantic

### PR DESCRIPTION
**Pull request by cristiancreteanu**
While trying to change the implementation of DCD to use DMD as a library, I have come across the need to retrieve the Scope of a Statement (namely the Statement at the location in the source file where the cursor is) without having too much of an impact on DMD's performance, but I couldn't find an easy way to do so by only using a visitor on the AST, as not all nodes had a Scope field or, if they did, it was null (in particular the _scope field of Dsymbol, which was null, except for field declarations). I took Basile's suggestions on the forum (https://forum.dlang.org/post/pinxxdclrvdnidpnxixr@forum.dlang.org).

With the current approach, however, I will not get results if the node whose Scope I am trying to find is a field declaration inside a class/struct/enum. To solve this problem, I could either use the same approach that I used for Statement and add delegates for Dsymbol in dsymbolsem or I could implement a custom visitor which traverses the AST and reaches the required node (which will be a Dsymbol whose _scope is set). What do you think is the best solution in this case?

### CODE REVIEW

**Comment by ibuclaw** [comment]
Would make more sense maybe if these could be fields inside the Compiler structure.

**Comment by rainers** [comment]
I have added a ScopeDsymbol to FuncDeclaration and ScopeStatement for a similar purpose, see https://github.com/rainers/dmd/commit/bc9b1c74fa77c7341434eeb66e19e37b5d0ad26a

The visitor to find the AST node for the caret is here: https://github.com/dlang/visuald/blob/master/vdc/dmdserver/semvisitor.d#L625 but there is also some additional machinery in place, e.g. remembering the original AST when it is modified during semantic analysis.

**Comment by cristiancreteanu** [comment]
> Would make more sense maybe if these could be fields inside the Compiler structure.

Now that you mention this, maybe keeping them on a global level isn't such a good idea, but I would rather add them as static fields inside the Statement class, as I am thinking that, since they are used for the semantic analysis of Statement instances, that's where they would probably be suited to be. But, based on your experience, could you please tell me why it would be better to add them in the Compiler structure?

**Comment by cristiancreteanu** [comment]
> I have added a ScopeDsymbol to FuncDeclaration and ScopeStatement for a similar purpose, see [rainers@bc9b1c7](https://github.com/rainers/dmd/commit/bc9b1c74fa77c7341434eeb66e19e37b5d0ad26a)
> 
> The visitor to find the AST node for the caret is here: https://github.com/dlang/visuald/blob/master/vdc/dmdserver/semvisitor.d#L625 but there is also some additional machinery in place, e.g. remembering the original AST when it is modified during semantic analysis.

Thanks! So I understand that VisualD uses your fork instead of the latest version of DMD. How come you haven't merged those changes to DMD? It would also solve my issue.

**Comment by rainers** [comment]
> Thanks! So I understand that VisualD uses your fork instead of the latest version of DMD.

Yes.



> How come you haven't merged those changes to DMD? It would also solve my issue.

Mostly to avoid endless discussions and fighting with the CI;-) Not all changes might be uncontroversial, and some probably need polishing.

https://github.com/dlang/dmd/pull/10846 was actually an attempt to start with creating small PRs, but that hit a road block because some tooling seems to depend on the current state, but it is outside  the dlang repositories.

**Comment by ibuclaw** [comment]
> > Would make more sense maybe if these could be fields inside the Compiler structure.
> 
> Now that you mention this, maybe keeping them on a global level isn't such a good idea, but I would rather add them as static fields inside the Statement class, as I am thinking that, since they are used for the semantic analysis of Statement instances, that's where they would probably be suited to be. But, based on your experience, could you please tell me why it would be better to add them in the Compiler structure?

There's already Compiler.onImport and Compiler.loadModule (which I want to get renamed to something like onModuleLoad) which have similar callbacks from frontend to compiler driver. This module is intended to be implemented independently by each glue frontend, so gdc and ldc can each implement their own actions if required.

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
Add a test and we are good to go.

Is anyone against this addition? This makes dmd as a library usable with DCD.

**Review by MoonlightSentinel — DISMISSED** [review]
Please also hide the other changes behind `version (Callback_API)`, they are unused when compiling without this version.

Otherwise LGTM

**Comment by cristiancreteanu** [comment]
I separated the creation of a single module so that I could use the function as part of the compiler's interface, as it can be seen in the test. It was also possible to use the already existent `createModules`, but it was a bit less straightfoward imo. Having the use case in the test in mind, I realised that if I wanted to work with a single module, first I'd have had to create an array of module file names with only one element, then call `createModules` and get the first element in the array of modules. This change also separates the creation of a module from the additional logic inside `createModules` and makes clear which part of the function handles the actual instantiation of a single module. Let me know if I should be working with the existing interface of mars.d instead and revert the change.

File: dub.sdl [hunk 1]
```diff
@@ -31,6 +31,8 @@ subPackage {
     "src/dmd/utf.d" \
     "src/dmd/utils.d"
 
+  versions "callback_API"
```
Comment (MoonlightSentinel) [inline dub.sdl:34]: I think this version should follow the usual style, `CallbackAPI` (preferred) or `Callback_API`

File: test/dub_package/retrieveScope.d [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        if (s.loc.linnum == cursorLoc.linnum) {
+            sc.setNoFree();
+            scp = sc;
+        }
+    }
+};
+
+int main()
+{
+    string dirName(string path, char separator)
```
Comment (MoonlightSentinel) [inline test/dub_package/retrieveScope.d:61]: You already use phobos in this test, so why not use `std.path.dirName`?

File: src/dmd/statementsem.d [hunk 3]
```diff
@@ -123,8 +124,17 @@ private Expression checkAssignmentAsCondition(Expression e)
 // Performs semantic analysis in Statement AST nodes
 extern(C++) Statement statementSemantic(Statement s, Scope* sc)
 {
+    version (callback_API)
+        if (Compiler.onStatementSemanticStart)
+            Compiler.onStatementSemanticStart(s, sc);
```
Comment (MoonlightSentinel) [inline src/dmd/statementsem.d:129]: Maybe give `onStatementSemantic(Start|End)` a no-op implementation by default to avoid these null checks?

**Review by RazvanN7 — APPROVED** [review]

File: src/dmd/mars.d [hunk 4]
```diff
@@ -2614,124 +2614,150 @@ private void reconcileCommands(ref Param params, size_t numSrcFiles)
...
+    version (Windows)
     {
-        const(char)[] name;
-        version (Windows)
+        file = toWinPath(file);
+    }
+    const(char)[] p = file.toDString();
+    p = FileName.name(p); // strip path
+    const(char)[] ext = FileName.ext(p);
+    if (ext)
```
Comment (thewilsonator) [inline src/dmd/mars.d:2642]: please invert the order of the order of these and early return of the (current) else branch.

Comment (cristiancreteanu) [reply src/dmd/mars.d:2642]: I am afraid I am not entirely sure what you're referring to. This is the exact workflow that `mars.d` had, I just extracted this from `createModules` and created a separate function.

Comment (thewilsonator) [reply src/dmd/mars.d:2642]: As in:
```
if (!ext)
{
     if (!p.length)
     {
           error(Loc.initial, "invalid file name '%s'", file);
           fatal();
     }
     auto id = Identifier.idPool(p);
     return new Module(file.toDString, id, global.params.doDocComments, global.params.doHdrGeneration);
}
...
```

**Review by MoonlightSentinel — COMMENTED** [review]
The internal changes are fine by me. Cannot really comment on it's usefullness for DMD as a library though.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**